File: ruby/os/linux/gentoo/glsa.rb
class thGlsa
	def initialize(user, host)
		if not user: @user = nil; end
		if not host: @host = nil; end
		@garg = ''
		@glsa = `which glsa-check`
		if not @glsa: puts 'Are you sure your on Gentoo?'; exit; end
		@applied = {}
		@unaffct = {}
		@affectd = {}
	end

	def addUnaffected(st)
		@unaffct[st.chomp.chop.split[0]] = st.chomp.chop.split(st.chomp.chop.split[1])[1].chomp
	end

	def addAffected(st)
		@affectd[st.chomp.chop.split[0]] = st.chomp.chop.split(st.chomp.chop.split[1])[1].chomp
	end

	def addApplied(st)
		@applied[st.chomp.chop.split[0]] = st.chomp.chop.split(st.chomp.chop.split[1])[1].chomp
	end

	def printStatus
		puts "Applied:( #{@applied.len} ) Unaffected:( #{@unaffct.len} ) Affected:( #{@affectd.len}\r"
		return
	end

	def check
		if @host == nil
			puts "Checking system security status ... \n"
		else
			puts "Checking system security status on #{@host} ... \n"
		end
		puts "Fetching current GLSA list ... \r"
		# register thSlurp() instance
		# register trigger
		# register trigger
		# register trigger
		if @host == nil and @user == nil
			# define process pipe object
		elsif @host != nil and @user != nil
			# define process pipe object
		else
			# define last hope pipe object
		end
		# slurp the damn pipe object
		puts "\n"
		@affectd.each { |pkg|
			puts "INFO: #{pkg} : #{@affectd[pkg]}\n" }
	end
end
		return
	end

File: python/THWAP/os/linux/gentoo/glsa.py
#!/usr/bin/env python
#
# Copyright (c) 2007 Academic Superstore
# Copyright (c) 2007 Mike "Fuzzy" Partin <fuzzy@academicsuperstore.com>
#
# Program code
import sys, os, popen2
from THWAP.core import config

class thGlsa:
	def __init__(self, user=None, host=None):
		self.host = host
		self.user = user
		self.garg = ''
		self.glsa = '/usr/bin/glsa-check'
		self.applied = {}
		self.unaffct = {}
		self.affectd = {}

	def addUnaffected(self, st=''):
		if st != '':
			tmp = st.strip().split()
			name = tmp[0]
			self.unaffct[name] = st.strip().split(st.strip().split()[1])[1].strip()
			self.printStatus()
		else:
			return False

	def addAffected(self, st=''):
		if st != '':
			tmp = st.strip().split()
			name = tmp[0]
			self.affectd[name] = st.strip().split(st.strip().split()[1])[1].strip()
			self.printStatus()
		else:
			return False

	def addApplied(self, st=''):
		if st != '':
			tmp = st.strip().split()
			name = tmp[0]
			self.applied[name] = st.strip().split(st.strip().split()[1])[1].strip()
			self.printStatus()
		else:
			return False

	def printStatus(self):
		sys.stdout.write('Applied:( %5d ) Unaffected:( %5d ) Affected:( %5d )\r' % (len(self.applied.keys()),len(self.unaffct.keys()),len(self.affectd.keys())))
		sys.stdout.flush()

	def check(self):
		if self.host == None:
			sys.stdout.write('Checking system security status ... \n')
		else:
			sys.stdout.write('Checking system security status on %s ... \n' % self.host)
		sys.stdout.write('Fetching current GLSA list...\r')
		sys.stdout.flush()
		self.slurp = config.thSlurp()
		self.slurp.registerTrigger('^[0-9]+.*\[U\]', self.addUnaffected)
		self.slurp.register_trigger('^[0-9]+.*\[A\]', self.addApplied)
		self.slurp.register_trigger('^[0-9]+.*\[N\]', self.addAffected)
		if self.host != None and self.user == None:
			obj = popen2.Popen4('ssh %s@%s "%s -ln"' % (os.getenv('USER'),self.host,self.glsa))
		elif self.host != None and self.user != None:
			obj = popen2.Popen4('ssh %s@%s "%s -nl"' % (self.user,self.host,self.glsa))
		else:
			obj = popen2.Popen4('%s -ln' % self.glsa)
		
		self.slurp.run(obj.fromchild)

		print ''
		for i in self.affectd.keys():
			sys.stdout.write('INFO: %s : %s\n' % (i, self.affectd[i]))

if __name__ == '__main__':
	o = GLSA()
	o.check()
